File: claude_exp/routers/router_cpp/src/router/crypto_client.cpp
#include "router/crypto_client.h"

#include <nlohmann/json.hpp>

#include "shared/base64.h"
#include "shared/log.h"

namespace router {

namespace {

// One worker thread's cached httplib::Client plus the bookkeeping that decides whether it's
// still usable - bundled into a single object instead of three loose thread_local variables kept
// in sync by hand, same idiom as this session's router_py/upstream_host refactor (_CachedConnection
// / _Connection: a cached resource that can go stale is modeled as one object with its own
// validity state, not scattered fields). httplib::Client isn't default-constructible without
// host/port, so this is heap-allocated (thread_local std::unique_ptr) rather than a bare
// thread_local struct.
struct CachedClient {
    httplib::Client client;
    int generation;
    std::chrono::steady_clock::time_point last_used;

    CachedClient(httplib::Client c, int gen)
        : client(std::move(c)), generation(gen), last_used(std::chrono::steady_clock::now()) {}
};

}  // namespace

CryptoClient::CryptoClient(const CryptoConfig& cfg, int breaker_threshold, int breaker_cooldown_seconds,
                           double idle_timeout_seconds)
    : cfg_(cfg),
      base_path_("/sys/v1/plugins/" + cfg.plugin_id),
      breaker_threshold_(breaker_threshold),
      breaker_cooldown_seconds_(breaker_cooldown_seconds),
      idle_timeout_seconds_(idle_timeout_seconds) {}

httplib::Client CryptoClient::make_client() const {
    if (cfg_.ssl_active) {
        std::string url = "https://" + cfg_.host + ":" + std::to_string(cfg_.port);
        // Same self-signed file doubles as this client's own identity (mutual TLS) when given -
        // matches router_py's crypto_client.py session.cert/verify split.
        httplib::Client client = (!cfg_.certfile.empty() && !cfg_.keyfile.empty())
                                      ? httplib::Client(url, cfg_.certfile, cfg_.keyfile)
                                      : httplib::Client(url);
        if (!cfg_.cafile.empty()) {
            client.set_ca_cert_path(cfg_.cafile);
            client.enable_server_certificate_verification(true);
        } else {
            client.enable_server_certificate_verification(false);
        }
        client.set_connection_timeout(5, 0);
        client.set_read_timeout(5, 0);
        client.set_write_timeout(5, 0);
        // httplib::Client defaults keep_alive_ to false, which closes (and for SSL, shuts down)
        // the socket after every single request regardless of this thread_local Client object
        // being reused across the thread's lifetime (see validate() below) - so without this,
        // every request pays a full mTLS handshake, not just the first one per thread. Matches
        // crypto_host_main.cpp's set_keep_alive_max_count(10000) on the server side, which only
        // helps if the client actually holds the connection open long enough to use it.
        client.set_keep_alive(true);
        client.set_tcp_nodelay(true);
        return client;
    }

    httplib::Client client(cfg_.host, cfg_.port);
    client.set_connection_timeout(5, 0);
    client.set_read_timeout(5, 0);
    client.set_write_timeout(5, 0);
    client.set_keep_alive(true);
    client.set_tcp_nodelay(true);
    return client;
}

void CryptoClient::record_failure() {
    std::lock_guard<std::mutex> lock(breaker_mutex_);
    ++failure_count_;
    if (failure_count_ >= breaker_threshold_) {
        open_until_ = std::chrono::steady_clock::now() + std::chrono::seconds(breaker_cooldown_seconds_);
        ++generation_;
    }
}

void CryptoClient::reset_failure() {
    std::lock_guard<std::mutex> lock(breaker_mutex_);
    failure_count_ = 0;
}

void CryptoClient::reset_breaker() {
    std::lock_guard<std::mutex> lock(breaker_mutex_);
    failure_count_ = 0;
    open_until_ = std::chrono::steady_clock::time_point{};
    ++generation_;
}

int CryptoClient::generation() {
    std::lock_guard<std::mutex> lock(breaker_mutex_);
    return generation_;
}

std::optional<std::string> CryptoClient::validate(const std::string& endpoint, const std::string& pan,
                                                   const std::string& f47, const std::string& router_stan) {
    {
        std::lock_guard<std::mutex> lock(breaker_mutex_);
        if (std::chrono::steady_clock::now() < open_until_) {
            return std::nullopt;
        }
    }

    nlohmann::json body = {{"operation", endpoint}, {"f2", pan}, {"f47", f47}, {"router_stan", router_stan}};
    httplib::Headers headers = {{"Authorization", "Bearer " + cfg_.bearer_token}};

    // One httplib::Client per calling thread (worker or response-worker), lazily built on
    // first use and reused for that thread's lifetime - httplib::Client isn't safe to share
    // across concurrent threads, and the dispatcher's worker pools call validate() from up to
    // worker_threads + response_worker_threads threads at once.
    //
    // Discarded and rebuilt proactively (not just on the next failed call) in two cases: (1) this
    // thread's cached client was built under a since-superseded generation - crypto_host died and
    // the breaker opened since, so the cached client is presumed stale even though this thread
    // itself hasn't seen a failure yet; (2) the cached client has been idle past
    // idle_timeout_seconds_ - crypto_host (cpp-httplib) closes its own keep-alive connections
    // after ~5s idle regardless of any breaker event, and a thread_local client sitting unused
    // across a traffic lull (e.g. a soak's tail-settle sleep between stress windows) has no way to
    // learn that from the client side short of trying it and failing. Same "anticipate staleness,
    // don't discover it by failing first" principle in both cases - mirrors router_py's
    // crypto_client.py _get_connection().
    thread_local std::unique_ptr<CachedClient> cached;
    int current_generation = generation();
    auto now = std::chrono::steady_clock::now();
    if (cached) {
        bool idle_too_long = (now - cached->last_used) > std::chrono::duration<double>(idle_timeout_seconds_);
        if (cached->generation != current_generation || idle_too_long) {
            cached.reset();
        }
    }
    if (!cached) {
        cached = std::make_unique<CachedClient>(make_client(), current_generation);
    } else {
        cached->last_used = now;
    }
    httplib::Client& client = cached->client;

    auto res = client.Post(base_path_, headers, body.dump(), "application/json");
    if (!res || res->status >= 400) {
        throttle_.log(shared::LogLevel::Warning, "crypto_call_failed:" + endpoint,
                       "crypto_client: " + endpoint + " request failed (router_stan=" + router_stan +
                           "), status=" + (res ? std::to_string(res->status) : std::string("(no response)")));
        record_failure();
        return std::nullopt;
    }

    try {
        // The 2XX body is a JSON string literal -- the PluginOutput envelope (base64,
        // "format": "byte"), not a JSON object wrapping it.
        std::string base64_str = nlohmann::json::parse(res->body).get<std::string>();
        auto decoded_bytes = shared::base64_decode(base64_str);
        std::string decoded_json(decoded_bytes.begin(), decoded_bytes.end());
        auto inner = nlohmann::json::parse(decoded_json);
        std::string f47_out = inner.value("f47", std::string(""));
        reset_failure();
        return f47_out;
    } catch (const std::exception& e) {
        throttle_.log(shared::LogLevel::Warning, "crypto_call_failed:" + endpoint,
                       "crypto_client: failed to decode PluginOutput envelope (router_stan=" + router_stan +
                           "): " + e.what());
        record_failure();
        return std::nullopt;
    }
}

}  // namespace router

File: claude_exp/routers/router_java/src/main/java/com/router/router/CryptoClient.java
package com.router.router;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.router.shared.LogThrottle;
import com.router.shared.SslUtils;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Base64;
import java.util.Map;
import java.util.concurrent.Semaphore;
import java.util.logging.Level;
import java.util.logging.Logger;

/** Fortanix-shaped crypto client: POST /sys/v1/plugins/{plugin_id}, bearer auth, base64 response.
 * Wired this way so swapping in a real Fortanix DSM tenant is a config/URL change, not a rewrite. */
public class CryptoClient {

    private static final Logger logger = Logger.getLogger(CryptoClient.class.getName());
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final LogThrottle throttledLog = new LogThrottle(logger, 200);

    private final String baseUrl;
    private final String bearerToken;
    private final int breakerThreshold;
    private final int breakerCooldownSeconds;
    private final long idleTimeoutMillis;
    private final Object lock = new Object();
    private int failureCount = 0;
    private long openUntilMillis = 0;
    // Bumped every time the breaker opens (and in resetBreaker()) - see validate()'s failure
    // branch. A thread's cached client is only trusted if it was built under the *current*
    // generation; ThreadLocal means one thread can't reach into another thread's cached client to
    // discard it directly when crypto_host dies, so instead every thread checks its own client's
    // generation the next time it needs one and discards it on mismatch, before ever trying to use
    // it - see getClient(). Never retried inline (see sendRequest()): a failed call is forgotten,
    // not rescued, on the theory that recovering *this* transaction isn't the goal - having a
    // fresh client ready for the next one is. Mirrors router_py's crypto_client.py _generation.
    private int generation = 0;

    // One HttpClient/SSLContext per dispatcher worker thread rather than one shared instance -
    // see build_router.md's CryptoClient section: a single HttpClient driven to open several
    // brand-new mTLS connections concurrently from different threads hits a genuine
    // java.net.http.HttpClient concurrency bug (intermittent "fatal alert: internal_error" on
    // ~1-2 of every 8 concurrent first-use connections, reproduced in isolation outside this
    // router entirely). Giving each thread its own fully independent client eliminated it in the
    // same isolated test - confirmed the shared instance, not crypto_host, was at fault.
    //
    // Bundled with its generation and last-used time into one CachedClient object rather than
    // three loose ThreadLocal fields kept in sync by hand - same idiom router_py's
    // crypto_client.py (_CachedConnection) and upstream_host/main.py (_Connection) use for the
    // analogous "a cached resource can go stale" problem. See getClient()/forgetClient().
    private final ThreadLocal<CachedClient> clientTL = new ThreadLocal<>();
    private final CryptoConfig cfg;

    private static final class CachedClient {
        final HttpClient client;
        final int generation;
        long lastUsedMillis;

        CachedClient(HttpClient client, int generation, long lastUsedMillis) {
            this.client = client;
            this.generation = generation;
            this.lastUsedMillis = lastUsedMillis;
        }
    }

    // Even fully independent HttpClient instances aren't safe to *construct and first-use*
    // concurrently (same isolated test: sequential construction, then concurrent use -> 0
    // failures; concurrent construction -> the same alert again, shared SSLContext or not). This
    // gate serializes only that one-time moment per thread (see warmup()) without blocking
    // Dispatcher.start() itself - an earlier attempt that made start() launch worker threads one
    // at a time (waiting on each thread's own warm-up before starting the next) fixed the alert
    // but delayed the router's upstream-accept readiness by ~(workerThreads +
    // responseWorkerThreads) * ~1-1.5s, long enough to blow past stress_run.sh's own connect
    // timeout - self-serializing here instead lets start() return immediately as before.
    private final Semaphore warmupGate = new Semaphore(1);

    public CryptoClient(CryptoConfig cfg, int breakerThreshold, int breakerCooldownSeconds) throws IOException {
        this(cfg, breakerThreshold, breakerCooldownSeconds, 4.0);
    }

    /** idleTimeoutSeconds: a thread's cached client is proactively discarded once idle past this
     * long (default 4.0s, under crypto_host's own ~5s keep-alive idle timeout) - separate from the
     * generation check above: a client can go stale from an ordinary traffic lull alone, with no
     * breaker event at all. Reproduced directly (router_py side, same shared crypto_host): after
     * any ~5s+ gap in traffic, the next reuse of an idle cached connection fails on both the
     * request and response legs - same "anticipate staleness, don't discover it by failing first"
     * principle as the generation check, just against a timer instead of an event. Mirrors
     * router_py's crypto_client.py idle_timeout_seconds. */
    public CryptoClient(CryptoConfig cfg, int breakerThreshold, int breakerCooldownSeconds,
            double idleTimeoutSeconds) throws IOException {
        String scheme = cfg.ssl_active() ? "https" : "http";
        this.baseUrl = scheme + "://" + cfg.host() + ":" + cfg.port() + "/sys/v1/plugins/" + cfg.plugin_id();
        this.bearerToken = cfg.bearer_token();
        this.breakerThreshold = breakerThreshold;
        this.breakerCooldownSeconds = breakerCooldownSeconds;
        this.idleTimeoutMillis = Math.round(idleTimeoutSeconds * 1000.0);
        this.cfg = cfg;

        if (cfg.ssl_active()) {
            // Fail fast on a bad cert/key/ca path here, at startup - otherwise this would only
            // surface on a worker thread's first real validate() call, indistinguishable from an
            // ordinary crypto_host failure instead of a startup misconfiguration.
            SslUtils.buildClientContext(cfg.certfile(), cfg.keyfile(), cfg.cafile());
        }
    }

    /** Returns this thread's cached client, discarding and rebuilding it first if it was built
     * under a stale generation or has sat idle too long - mirrors router_py's
     * crypto_client.py _get_connection(). */
    private HttpClient getClient() {
        int currentGeneration;
        synchronized (lock) {
            currentGeneration = generation;
        }
        long now = System.currentTimeMillis();
        CachedClient cached = clientTL.get();
        if (cached != null) {
            boolean idleTooLong = (now - cached.lastUsedMillis) > idleTimeoutMillis;
            if (cached.generation != currentGeneration || idleTooLong) {
                closeQuietly(cached.client);
                cached = null;
            }
        }
        if (cached == null) {
            cached = new CachedClient(buildClient(cfg), currentGeneration, now);
            clientTL.set(cached);
        } else {
            cached.lastUsedMillis = now;
        }
        return cached.client;
    }

    /** Discards this thread's cached client outright - called after a genuine transport failure
     * (see sendRequest()) so the next call, this thread's or another's, starts from a known-fresh
     * client instead of inheriting a guessing game. Mirrors _forget_connection(). */
    private void forgetClient() {
        CachedClient cached = clientTL.get();
        if (cached != null) {
            closeQuietly(cached.client);
        }
        clientTL.remove();
    }

    private static void closeQuietly(HttpClient client) {
        try {
            client.close();
        } catch (Exception e) {
            // best-effort - being discarded either way
        }
    }

    /** Sends one request over this thread's cached client, forgetting (closing + discarding) that
     * client on any transport-level failure - never retried inline (see class-level generation
     * doc). Deliberately does NOT forget the client on an HTTP-level error status (>=400, checked
     * by the caller after this returns) - a bad status over a working connection isn't evidence
     * the connection itself is broken. Mirrors _send(). */
    private HttpResponse<String> sendRequest(HttpRequest request) throws IOException, InterruptedException {
        HttpClient client = getClient();
        try {
            return client.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (IOException | InterruptedException e) {
            forgetClient();
            throw e;
        }
    }

    private static HttpClient buildClient(CryptoConfig cfg) {
        HttpClient.Builder builder = HttpClient.newBuilder();
        if (cfg.ssl_active()) {
            try {
                builder.sslContext(SslUtils.buildClientContext(cfg.certfile(), cfg.keyfile(), cfg.cafile()));
            } catch (IOException e) {
                // Already validated once in the constructor - a failure here means the cert files
                // changed or became unreadable after startup, not a config typo.
                throw new UncheckedIOException(e);
            }
        }
        return builder.build();
    }

    /**
     * Returns the enriched f47 on success, or "" on any failure (breaker open or HTTP error) -
     * callers only overwrite their working f47 when this return value is non-empty, so any
     * failure path leaves the original f47 untouched. Handles Fortanix PluginOutput envelope:
     * response body is a JSON string literal (base64 "format":"byte"), which we decode to reach
     * the inner {"f47": ...} object.
     *
     * routerStan isn't part of the Fortanix plugin contract - it's passed through so crypto_host's
     * own logs can be joined with this router's logs on the same transaction (mirrors router_py's
     * crypto_client.py). Empty string when a caller has none.
     */
    private static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(5);
    // Cold JVM TLS start (KeyManagerFactory/TrustManagerFactory class-loading and JIT, cipher
    // negotiation) has been measured up to ~4.6s in isolation on an idle host, and observed
    // exceeding the standard 5s REQUEST_TIMEOUT outright on this host under real load (every
    // worker/response-worker thread's first-ever call, serialized through warmupGate, hitting
    // "HTTP connect timed out" back to back) - a longer timeout only for that one-time first call
    // gives it enough room without loosening the 5s deadline real (already-warm) traffic is held
    // to, where a slow crypto_host is a genuine problem worth detecting quickly.
    private static final Duration WARMUP_TIMEOUT = Duration.ofSeconds(20);

    /** Returns the enriched f47 on success (possibly "" if crypto_host genuinely had nothing to
     * add), or null on any genuine failure (breaker open or HTTP error) - mirrors router_py's
     * crypto_client.py (None vs "") and router_cpp's CryptoClient (std::nullopt vs ""). A caller
     * that only needs "fail open" (the request leg) can keep treating both null and "" as "don't
     * overwrite"; a caller that must distinguish "no-op success" from "genuine failure" (the
     * response leg, which needs to drop rather than forward unvalidated) must check for null
     * specifically, not just isEmpty(). */
    public String validate(String endpoint, String pan, String f47, String routerStan) {
        return validate(endpoint, pan, f47, routerStan, REQUEST_TIMEOUT);
    }

    private String validate(String endpoint, String pan, String f47, String routerStan, Duration timeout) {
        synchronized (lock) {
            if (System.currentTimeMillis() < openUntilMillis) {
                return null;
            }
        }

        try {
            String json = MAPPER.writeValueAsString(
                    Map.of("operation", endpoint, "f2", pan, "f47", f47, "router_stan", routerStan));
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl))
                    .timeout(timeout)
                    .header("Content-Type", "application/json")
                    .header("Authorization", "Bearer " + bearerToken)
                    .POST(HttpRequest.BodyPublishers.ofString(json))
                    .build();
            HttpResponse<String> response = sendRequest(request);
            if (response.statusCode() >= 400) {
                throw new IOException("HTTP " + response.statusCode());
            }

            String body = response.body();
            if (body == null || body.isEmpty()) {
                throw new IOException("empty response body");
            }

            String base64String = MAPPER.readValue(body, String.class);
            String decodedJson = new String(Base64.getDecoder().decode(base64String), StandardCharsets.UTF_8);
            Map<?, ?> parsed = MAPPER.readValue(decodedJson, Map.class);
            Object result = parsed.get("f47");

            synchronized (lock) {
                failureCount = 0;
            }
            return result == null ? "" : result.toString();
        } catch (Exception e) {
            throttledLog.log(Level.WARNING, "crypto_call_failed:" + endpoint,
                    "crypto_host " + endpoint + " call failed (router_stan=" + routerStan
                            + "): " + e.getMessage());
            synchronized (lock) {
                failureCount++;
                if (failureCount >= breakerThreshold) {
                    openUntilMillis = System.currentTimeMillis() + breakerCooldownSeconds * 1000L;
                    generation++;
                    logger.warning("crypto breaker open for " + breakerCooldownSeconds
                            + "s after " + failureCount + " consecutive failures");
                }
            }
            return null;
        }
    }

    /** Closes the breaker immediately rather than waiting for its own cooldown clock to expire and
     * self-renew (see validate()'s failure branch, which re-arms openUntilMillis on every failed
     * post-cooldown probe with no awareness of whether the service actually recovered) - for a
     * caller that has *externally confirmed* the service is back (e.g. a TCP probe against its
     * port) and wants the breaker to trust that immediately instead of on its own delayed
     * schedule. Bumps generation like a normal open/close cycle so every thread's cached client is
     * rebuilt fresh, matching a real recovery. Mirrors router_py's crypto_client.py
     * reset_breaker(). */
    public void resetBreaker() {
        synchronized (lock) {
            failureCount = 0;
            openUntilMillis = 0;
            generation++;
        }
    }

    /** Forces the calling thread's own HttpClient/SSLContext (see clientTL above) through its
     * one-time cold-start cost (KeyManagerFactory/TrustManagerFactory class-loading and JIT,
     * cipher-suite negotiation) here, synchronously, before that thread starts handling real
     * traffic. Measured in isolation at ~0.5-1.5s on this host for the client build, then another
     * ~1-2s for the first request. Must be called once from each dispatcher worker/response-worker
     * thread itself (see Dispatcher.workerLoop/responseWorkerLoop) - calling it from any other
     * thread only warms that other thread's own ThreadLocal client, not the worker's.
     * {@code warmupGate} serializes this specific call across threads (see its own doc) so no two
     * threads' client construction/first-use ever overlaps, without blocking anything else. The
     * call's own success/failure isn't reported to the caller - even a failed warm-up call has
     * already paid the cold-start cost for whatever comes next on this thread. */
    public void warmup() {
        try {
            warmupGate.acquire();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return;
        }
        try {
            validate("validate_0100", "0", "", "warmup", WARMUP_TIMEOUT);
        } finally {
            warmupGate.release();
        }
    }
}
